cpmpy/solvers/cpo.py

```diff
@@ -0,0 +1,470 @@
+    Interface to CP Optimizers API
+
+    CP Optimizer, also a feature of IBM ILOG Optimization Studio, is a software library of constraint programming tools 
+    supporting constraint propagation, domain reduction, and highly optimized solution search.
+
+    Documentation of the solver's own Python API: (all modeling functions)
+    https://ibmdecisionoptimization.github.io/docplex-doc/cp/docplex.cp.modeler.py.html#module-docplex.cp.modeler
+
+    ===============
+    List of classes
+    ===============
+
+    .. autosummary::
+        :nosignatures:
+
+        CPM_cpo
+"""
+
+class CPM_cpo(SolverInterface):
+    """
+    Interface to CP Optimizers API
+
+    Requires that the 'docplex' python package is installed:
+    $ pip install docplex
+
+    docplex documentation:
+    https://ibmdecisionoptimization.github.io/docplex-doc/
+
+    You will also need to install CPLEX Optimization Studio from IBM's website.
+    There is a free community version available.
+    https://www.ibm.com/products/ilog-cplex-optimization-studio
+
+    See detailed installation instructions at:
+    https://www.ibm.com/docs/en/icos/22.1.2?topic=2212-installing-cplex-optimization-studio
+
+    Creates the following attributes (see parent constructor for more):
+    - cpo_model: object, CP Optimizers model object
+    """
+
+    _docp = None  # Static attribute to hold the docplex.cp module
+
+    @classmethod
+    def get_docp(cls):
+        if cls._docp is None:
+            import docplex.cp as docp  # Import only once
+            cls._docp = docp
+        return cls._docp
+
+    @staticmethod
+    def supported():
+        return CPM_cpo.installed() and CPM_cpo.license_ok()
+
+    @staticmethod
+    def installed():
+        # try to import the package
+        try:
+            import docplex.cp as docp
+            return True
+        except ModuleNotFoundError:
+            return False
+
+    @staticmethod
+    def license_ok():
+        if not CPM_cpo.installed():
+            warnings.warn(f"License check failed, python package 'docplex' is not installed! Please check 'CPM_cpo.installed()' before attempting to check license.")
+            return False
+        else:
+            try:
+                from docplex.cp.model import CpoModel
+                mdl = CpoModel()
+                mdl.solve(LogVerbosity='Quiet')
+                return True
+            except:
+                return False
+
+    def __init__(self, cpm_model=None, subsolver=None):
+        """
+        Constructor of the native solver object
+
+        Arguments:
+        - cpm_model: Model(), a CPMpy Model() (optional)
+        - subsolver: str, name of a subsolver (optional)
+        """
+        if not self.installed():
+            raise Exception("CPM_cpo: Install the python package 'docplex'")
+
+        if not self.license_ok():
+            raise Exception("You need to install the CPLEX Optimization Studio to use this solver.")
+
+        import docplex.cp.model as dom
```

Comment:
this also self._docp...?

Reply:
ah yes I thought it didn't work within init but it was probably the static methods that didn't work